### PR TITLE
Add support for pedantic property access

### LINKED ISSUE #40170: pedanticPropertyLookup
## Search Terms
type check

## Suggestion

As https://gist.github.com/RyanCavanaugh/f80f9ddc50d45c4d76e7c4101efada28 said.

> To turn off allowing `expr.prop` when expr just has a string index signature.

But does it's better if only issue errors when property access referenced to signature?

## Examples

```ts
interface A {
    foo: string
}
interface B {
    [k: string]: string
}
interface C {
    foo: string
    [k: string]: string
}
declare const a: A;
declare const b: B;
declare const c: C;
declare const d: C | undefined;
a.foo;
a["foo"]
b.foo; // error
b["foo"];
c.foo;
c["foo"]
c.bar; // error
c["bar"];
d?.foo;
d?.["foo"]
d?.bar; // error
d?.["bar"];
```

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [ ] This wouldn't change the runtime behavior of existing JavaScript code
* [ ] This could be implemented without emitting different JS based on the types of the expressions
* [ ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [ ] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #40170

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -3386,6 +3386,10 @@
         "category": "Error",
         "code": 4110
     },
+    "Access property '{0}' by index signature is disallowed.": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json:3389]: Property '{0}' comes from an index signature, so it must be accessed with ["{0}"].
OR
"... must be accessed with element access."

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -4686,6 +4690,10 @@
         "category": "Error",
         "code": 6504
     },
+    "Force access member from index signature with element access expression.": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json:4693]: Require undeclared properties from index signatures to use element accesses.

File: src/compiler/diagnosticMessages.json [hunk 3]
```diff
@@ -5876,6 +5884,14 @@
         "category": "Message",
         "code": 95142
     },
+    "Use element access for '{0}'": {
+        "category": "Message",
+        "code": 95143
+    },
+    "Use element access for all property access": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json:5891]:
```suggestion
    "Use element access for all undeclared properties.": {

```

**Review by sandersn — CHANGES_REQUESTED** [review]
Some suggestions for the wording of messages.

**Comment by sandersn** [comment]
Also, at the Friday (Sep 4) design meeting, we decided not to create a single naming scheme for pedantic options, but (1) to name them for what they do (2) use the "no-" prefix only if it makes sense, since we already have some pedantic options named that, like noUnusedLocals, noFallthroughCasesInSwitch, etc.

I would suggest noPropertyAccessFromIndexSignature.

**Review by andrewbranch — APPROVED** [review]
Implementation looks good, @sandersn’s messaging suggestions look good

**Comment by ssledorze** [comment]
Really useful and awaited! :)
Looking forward to that PR being merged..
Is there some information on the expected release it would be part of?

**Comment by Kingwl** [comment]
@ssledorze After 4.1 release, I guess.

**Review by sandersn — APPROVED** [review]
I'll merge this after 4.1 RC.

**Comment by ssledorze** [comment]
> I'll merge this after 4.1 RC.

should I understand it will be in the 4.1 release @sandersn  ?

**Comment by sandersn** [comment]
No, we're finishing up 4.1 RC right now, then 4.1 will get its own branch. After that, commits to master will go into 4.2. So this will be in 4.2.

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Pull request merged**